dive_core/data_core.h

```diff
@@ -173,4 +176,42 @@ class CaptureMetadataCreator : public IEmulateCallbacks
 #endif
 };
 
+//--------------------------------------------------------------------------------------------------
+// validate LRZ from the capture
+class LRZValidator : public IEmulateCallbacks
```

Comment:
hmm, maybe i don't fully understand your idea. I put some extra logic in the OnPacket() function which are not part of the statetracker. (trying to separate the validation output code from the state tracker)

Reply:
The idea is that you rely on the existing CaptureMetadataCreator. Have it fill out all the EventStateInfo, as it usually does for Dive. Then in your main.cpp you can simply loop through the EventStateInfo and do your depth_test and lrz state logic. 

The idea is that one-off CLI commands can rely all the heavy lifting code-wise that we already wrote. They can just look at the metadata we determined rather than re-write callbacks and handle parsing of PM4 packets and duplicate some logic. If we're missing anything in CaptureMetadataCreator that blocks your attempt to do this, then maybe that just means we're missing stuff in CaptureMetadataCreator?

Reply:
In other words, I would've imgained your lrz_validator to just use dive_core as-is without modifications, rather than require actually making changes in dive_core. Any changes that we add to dive_core should be minimal and indicative of information we didn't parse or cache properly to begin with. In principle anyways.

Reply:
I see your point. My worries is that CaptureMetadataCreator is doing lots of other things that are not requried by the lrz_validator and will slowdown the lrz_validator (it is currently already kinda slow, for the firsthand scene with ~100 drawcalls, it is taking ~30s) The `EventStateInfo` contains the states for all drawcalls (binning + multiple tiles) whereas the lrz_validator only wants to show the drawcalls in the binning pass (we could add an extra field in EventStateInfo to indicate if the drawcall is in the binning pass, a bit weird if it is only used for lrz_validator)

Reply:
I would be surprised if there's a real slowdown in CaptureMetadataCreator, because it's something we run everytime we load a capture. So far that never takes ~30s, and if it does, then it means we need to optimize it. I'd be curious what's taking so long for lrz_validator.

Yes, EventStateInfo will always contain more info than you could possibly need, but that's why it's so broadly useful. We can create lots of one-off CLI tools without worrying about parsing, because CaptureMetadataCreator should meet most needs. I can imagine other clients of dive_core (including Dive itself!) might want to differentiate binning vs tile passes, so it seems reasonable to add that info to CaptureMetadataCreator.

Reply:
currently, ~90% of the time is spent in the loading of the texture, and the loading lrz function is taking ~7.5%  which is ~2.3s. (half of it are spent in Dive::EmulatePM4::AdvanceCb(Dive::MemoryManager::GetMaxContiguousSize) and the other half are spent in Dive::MemoryManager::CopyMemory)

Another thing i am worried is that one of the optimizations that i am thinking of is to merge the `create trees` with the `metatdata creator` so that we don't need to loop through the entire packets 2 times (we could avoid those heavy memory operations), but that also means we pay the cost of creating tress for lrz cli as well. 

in the latest commit, to avoid polluting dive_core, i moved lrz to main.cpp, so we do all the validation logic in the cli project. What do you think?

Reply:
Updated the code as we discussed. The 30+sec is an old capture with a size of 1GB (~18GB after decompression), this is without robin's fix. With [robin's fix](https://github.com/google/dive/pull/87), the size of the capture from the same scene is about ~160MB (compressed), and loading takes about 6s.